### ISSUE TITLE
`LogicException` when running `composer init` and specifying requires interactively

### ISSUE THREAD

**Issue by dwminer** [composer/composer#10739]
**My `composer.json`:**

N/A

**Output of `composer diagnose`:**

```
dminer $ composer diagnose
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Composer version: 2.3.5
PHP version: 8.0.17
PHP binary path: /usr/bin/php
OpenSSL version: OpenSSL 1.1.1l  FIPS 24 Aug 2021
cURL version: 7.79.1 libz 1.2.11 ssl OpenSSL/1.1.1n-fips
zip: extension present, unzip present, 7-Zip not available
```

**When I run this command:** 

```
composer init -vvv
```

**I get the following output:** 

```
Running 2.3.5 (2022-04-13 16:43:00) with PHP 8.0.17 on Linux / 5.16.18-200.fc35.x86_64
Loading config file /home/dminer/.composer/config.json
Loading config file /home/dminer/.composer/auth.json
Failed to initialize global composer: Composer could not find the config file: /home/dminer/.composer/composer.json
To initialize a project, please create a composer.json file. See https://getcomposer.org/basic-usage

                                            
  Welcome to the Composer config generator  
                                            


This command will guide you through creating your composer.json config.

Package name (<vendor>/<name>) [dminer/tmp]: 
Description []: 
Author [Dalton Miner <dminer@datto.com>, n to skip]: 
Minimum Stability []: 
Package Type (e.g. library, project, metapackage, composer-plugin) []: 
License []: 

Define your dependencies.

Loading config file /home/dminer/.composer/config.json
Loading config file /home/dminer/.composer/auth.json
Checked CA file /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem: valid
Would you like to define your dependencies (require) interactively [yes]? 
Search for a package: symfony/symfony
Reading /home/dminer/.composer/cache/repo/https---repo.packagist.org/packages.json from cache
Downloading https://packagist.org/search.json?q=symfony%2Fsymfony&type=
[200] https://packagist.org/search.json?q=symfony%2Fsymfony&type=
Enter the version constraint to require (or leave blank to use the latest version): 

In BaseCommand.php line 304:
                                                                                                                                  
  [LogicException]                                                                                                                
  Calling getPlatformRequirementFilter from a command which does not define the --ignore-platform-req[s] flags is not permitted.  
                                                                                                                                  

Exception trace:
  at /usr/share/php/Composer/Command/BaseCommand.php:304
 Composer\Command\BaseCommand->getPlatformRequirementFilter() at /usr/share/php/Composer/Command/PackageDiscoveryTrait.php:274
 Composer\Command\InitCommand->findBestVersionAndNameForPackage() at /usr/share/php/Composer/Command/PackageDiscoveryTrait.php:241
 Composer\Command\InitCommand->determineRequirements() at /usr/share/php/Composer/Command/InitCommand.php:415
 Composer\Command\InitCommand->interact() at /usr/share/php/Composer/vendor/symfony/console/Command/Command.php:283
 Symfony\Component\Console\Command\Command->run() at /usr/share/php/Composer/vendor/symfony/console/Application.php:1015
 Symfony\Component\Console\Application->doRunCommand() at /usr/share/php/Composer/vendor/symfony/console/Application.php:299
 Symfony\Component\Console\Application->doRun() at /usr/share/php/Composer/Console/Application.php:334
 Composer\Console\Application->doRun() at /usr/share/php/Composer/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at /usr/share/php/Composer/Console/Application.php:130
 Composer\Console\Application->run() at /usr/bin/composer:89

init [--name NAME] [--description DESCRIPTION] [--author AUTHOR] [--type [TYPE]] [--homepage HOMEPAGE] [--require REQUIRE] [--require-dev REQUIRE-DEV] [-s|--stability STABILITY] [-l|--license LICENSE] [--repository REPOSITORY] [-a|--autoload AUTOLOAD]
```

**And I expected this to happen:**
I would expect the command to select the latest available version, per the prompt. This does _not_ occur if you specify a version constraint interactively or via the `--require` option.

The prior version available in my distribution is composer-2.2.9 which does not have this problem, so it appears to be a regression in a recent version. As of this report, 2.3.5 is still in testing in my distribution, (Fedora 35, see https://bodhi.fedoraproject.org/updates/FEDORA-2022-47d2e7da46) but I can also reproduce this with composer installed with the instructions from https://getcomposer.org/download/. This applies to my base system as well as in a fresh `fedora:35` container.

**Comment by Lreus** [comment]
I confirm the problem occurs with composer version 2.3.5

**Comment by Cyclonecode** [comment]
I can also confirm that this is an issue with version 2.3.5 running PHP 8.1, not sure why this is closed? It does not seem to be fixed?

**Comment by Seldaek** [comment]
It's closed because 2.3.6 fixes it?

**Comment by philip** [comment]
Confirmed that it was fixed as per the [2.3.6 release notes](https://github.com/composer/composer/releases/tag/2.3.6) and from personal experience after stumbling into this issue and upgrading :)
